src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +24,181 @@
+    # This permits a non-fatal `kill -31 <pants pid>` for stacktrace retrieval.
+    faulthandler.register(signal.SIGUSR2, interactive_output_stream, chain=True)
+
+    # NB: mutate the class variables!
+    # We don't *necessarily* need to keep a reference to this, but we do here for clarity.
+    cls._interactive_output_stream = interactive_output_stream
 
   @classmethod
-  def exceptions_log_path(cls, for_pid=None):
-    intermediate_filename_component = '.{}'.format(for_pid) if for_pid else ''
+  def exceptions_log_path(cls, for_pid=None, in_dir=None):
+    """Get the path to either the shared or pid-specific fatal errors log file."""
+    if for_pid is None:
+      intermediate_filename_component = ''
+    else:
+      assert(isinstance(for_pid, IntegerForPid))
+      intermediate_filename_component = '.{}'.format(for_pid)
+    in_dir = in_dir or cls._log_dir
     return os.path.join(
-      cls.get_destination(),
+      in_dir,
       'logs',
       'exceptions{}.log'.format(intermediate_filename_component))
 
+  @classmethod
+  def log_exception(cls, msg):
+    """Try to log an error message to this process's error log and the shared error log.
+
+    NB: Doesn't raise (logs an error instead).
+    """
+    pid = os.getpid()
+    fatal_error_log_entry = cls._format_exception_message(msg, pid)
+
+    # We care more about this log than the shared log, so write to it first.
+    try:
+      cls._try_write_with_flush(cls._pid_specific_error_fileobj, fatal_error_log_entry)
+    except Exception as e:
+      logger.error(
+        "Error logging the message '{}' to the pid-specific file handle for {} at pid {}:\n{}"
+        .format(msg, cls._log_dir, pid, e))
+
+    # Write to the shared log.
+    try:
+      # TODO: we should probably guard this against concurrent modification by other pants
+      # subprocesses somehow.
+      cls._try_write_with_flush(cls._shared_error_fileobj, fatal_error_log_entry)
+    except Exception as e:
+      logger.error(
+        "Error logging the message '{}' to the shared file handle for {} at pid {}:\n{}"
+        .format(msg, cls._log_dir, pid, e))
+
+  @classmethod
+  def _try_write_with_flush(cls, fileobj, payload):
+    """This method is here so that it can be patched to simulate write errors.
+
+    This is because mock can't patch primitive objects like file objects.
+    """
+    fileobj.write(payload)
+    fileobj.flush()
+
+  @classmethod
+  def _check_or_create_new_destination(cls, destination):
+    try:
+      safe_mkdir(destination)
+    except Exception as e:
+      raise cls.ExceptionSinkError(
+        "The provided exception sink path at '{}' is not writable or could not be created: {}."
+        .format(destination, str(e)),
+        e)
+
+  @classmethod
+  def _recapture_fatal_error_log_streams(cls, new_log_location):
+    # NB: We do not close old file descriptors! This is bounded by the number of times any method is
```

Comment:
I think that if we uninstall old signals, we should be able to do this? But maybe ok to do in a followup.

Reply:
I think this sounds like it jives with the other elements of a followup issue we've identified (moving streams into the Exiter, for example). Will make one.

Reply:
Broke out #6596!